eth2/utils/tree_hash/src/merkleize_padded.rs

```diff
@@ -0,0 +1,354 @@
+                (Err(_), Err(_)) => unreachable!("Parent must have one child"),
+                // `chunks` is a contiguous array so it is impossible for an index to be missing
+                // when a higher index is present.
+                (Err(_), Ok(_)) => unreachable!("Parent must have a left child"),
+            };
+
+            assert!(
+                left.len() == right.len() && right.len() == BYTES_PER_CHUNK,
+                "Both children should be `BYTES_PER_CHUNK` bytes."
+            );
+
+            let hash = hash_concat(&left.to_vec(), &right.to_vec());
+
+            // Store a parent node.
+            chunks
+                .set(i, &hash)
+                .expect("Buf is adequate size for parent");
+        }
+
+        // Shrink the buffer so it neatly fits the number of new nodes created in this round.
+        //
+        // The number of `parent_nodes` is either decreasing or stable. It never increases.
+        chunks.truncate(parent_nodes);
+    }
+
+    // There should be a single chunk left in the buffer and it is the Merkle root.
+    let root = chunks.into_vec();
+
+    assert_eq!(root.len(), BYTES_PER_CHUNK, "Only one chunk should remain");
+
+    root
+}
+
+/// A helper struct for storing words of `BYTES_PER_CHUNK` size in a flat byte array.
+#[derive(Debug)]
+struct ChunkStore(Vec<u8>);
+
+impl ChunkStore {
+    /// Creates a new instance with `chunks` padding nodes.
+    fn with_capacity(chunks: usize) -> Self {
+        Self(vec![0; chunks * BYTES_PER_CHUNK])
+    }
+
+    /// Set the `i`th chunk to `value`.
+    ///
+    /// Returns `Err` if `value.len() != BYTES_PER_CHUNK` or `i` is out-of-bounds.
+    fn set(&mut self, i: usize, value: &[u8]) -> Result<(), ()> {
+        if i < self.len() && value.len() == BYTES_PER_CHUNK {
+            let slice = &mut self.0[i * BYTES_PER_CHUNK..i * BYTES_PER_CHUNK + BYTES_PER_CHUNK];
+            slice.copy_from_slice(value);
+            Ok(())
+        } else {
+            Err(())
+        }
+    }
+
+    /// Gets the `i`th chunk.
+    ///
+    /// Returns `Err` if `i` is out-of-bounds.
+    fn get(&self, i: usize) -> Result<&[u8], ()> {
+        if i < self.len() {
+            Ok(&self.0[i * BYTES_PER_CHUNK..i * BYTES_PER_CHUNK + BYTES_PER_CHUNK])
+        } else {
+            Err(())
+        }
+    }
+
+    /// Returns the number of chunks presently stored in `self`.
+    fn len(&self) -> usize {
+        self.0.len() / BYTES_PER_CHUNK
+    }
+
+    /// Truncates 'self' to `num_chunks` chunks.
+    ///
+    /// Functionally identical to `Vec::truncate`.
+    fn truncate(&mut self, num_chunks: usize) {
+        self.0.truncate(num_chunks * BYTES_PER_CHUNK)
+    }
+
+    /// Consumes `self`, returning the underlying byte array.
+    fn into_vec(self) -> Vec<u8> {
+        self.0
+    }
+}
+
+/// Returns a cached padding node for a given height.
+fn get_zero_hash(height: usize) -> &'static [u8] {
+    if height < MAX_TREE_DEPTH {
```

Comment:
This can be `height <= MAX_TREE_DEPTH`, because the hash cache stores `MAX_TREE_DEPTH` + 1 hashes in order to be able to retrieve the root of an entirely 0 tree (this was useful in my old `MerkleTree` impl, maybe not so much here).

It makes this test viable though:

```rust
    #[test]
    fn max_tree_depth_min_nodes() {
        let input = vec![0; 10 * BYTES_PER_CHUNK];
        let min_nodes = 2usize.pow(MAX_TREE_DEPTH as u32);
        assert_eq!(
            merkleize_padded(&input, min_nodes),
            get_zero_hash(MAX_TREE_DEPTH)
        );
    }
```

Reply:
Resolved!